iac-extensions/docker/src/test/java/org/sonar/iac/docker/checks/PrivilegedUserCheckTest.java

```diff
@@ -19,20 +19,72 @@
  */
 package org.sonar.iac.docker.checks;
 
+import java.util.List;
+import java.util.function.BiConsumer;
 import org.junit.jupiter.api.Test;
+import org.mockito.Mockito;
+import org.sonar.api.batch.fs.TextRange;
+import org.sonar.iac.common.api.checks.CheckContext;
+import org.sonar.iac.common.api.checks.InitContext;
+import org.sonar.iac.common.api.tree.Tree;
+import org.sonar.iac.common.api.tree.impl.TextRanges;
+import org.sonar.iac.docker.tree.api.DockerImageTree;
+import org.sonar.iac.docker.tree.api.FromTree;
+import org.sonar.iac.docker.tree.api.SyntaxToken;
+import org.sonar.iac.docker.tree.impl.DockerImageTreeImpl;
+import org.sonar.iac.docker.tree.impl.FromTreeImpl;
+import org.sonar.iac.docker.tree.impl.ImageTreeImpl;
+import org.sonar.iac.docker.tree.impl.SyntaxTokenImpl;
 
 class PrivilegedUserCheckTest {
 
   private final PrivilegedUserCheck check = new PrivilegedUserCheck();
 
   @Test
-  void test() {
-    DockerVerifier.verify("PrivilegedUserCheck/Dockerfile", check);
+  void testNonCompliant() {
+    for (int i = 0; i < 10; i++) {
+      DockerVerifier.verify("PrivilegedUserCheck/Dockerfile." + i, check);
+    }
   }
 
   @Test
-  void test_customSafeList() {
+  void testCompliant() {
+    for (int i = 0; i < 7; i++) {
+      DockerVerifier.verifyNoIssue("PrivilegedUserCheck/Dockerfile-Compliant." + i, check);
+    }
+  }
+
+  @Test
+  void testCustomSafeList() {
+    check.safeImages = "custom_image1, custom_image2, golang";
+    for (int i = 0; i < 3; i++) {
+      DockerVerifier.verify("PrivilegedUserCheck/Dockerfile_customSafeImages." + i, check);
+    }
+  }
+
+  @Test
+  void testCustomSafeListCompliant() {
     check.safeImages = "custom_image1, custom_image2, golang";
-    DockerVerifier.verify("PrivilegedUserCheck/Dockerfile_customSafeImages", check);
+    for (int i = 0; i < 7; i++) {
+      DockerVerifier.verifyNoIssue("PrivilegedUserCheck/Dockerfile_customSafeImages-Compliant." + i, check);
+    }
+  }
+
+  @Test
+  void testMultiStageBuild() {
+    DockerVerifier.verify("PrivilegedUserCheck/Dockerfile_multi_stage_build", check);
+  }
+
+  @Test
+  void shouldRiseIssueWhenFileTreeParentIsNull() {
+    CheckContext ctx = Mockito.mock(CheckContext.class);
+    SyntaxToken imageName = new SyntaxTokenImpl("scratch", TextRanges.range(1,5, "scratch"), List.of());
+    SyntaxToken fromToken = new SyntaxTokenImpl("FROM", TextRanges.range(1,0, "FROM"), List.of());
+    FromTree from = new FromTreeImpl(fromToken, null, new ImageTreeImpl(imageName, null, null), null);
+    DockerImageTree tree = new DockerImageTreeImpl(from, List.of());
+
+    check.handle(ctx, tree);
+
+    Mockito.verify(ctx).reportIssue(from, "Scratch images run as root by default. Make sure it is safe here.");
   }
```

Comment:
Is this a plausible use case ? I find it strange that a `DockerImage` can have no parent.
The fact that you had to use a Mockito to simulate that tend to prove that we normally can't encounter such use case ^^'
Unless we have good reason for this, I would tend to remove this test + remove the condition `if (parent instanceof FileTree) {` in `isLastDockerImageInFile` and cast directly.

Reply:
Good point. `DockerImage` should always have a parent and it will be always `FileTree`. Unfortunately, Java type system (or our design) allows any `DockerTree` to be a parent. My intention when I wrote `if (parent instanceof FileTree)` was to not fail with ugly `ClassCastException`. However, the structure of the trees is in our control, so such exception means an issue in the parsing phase. 
So now, I'm ok with remove of the `if` and artificial test just for coverage.